### PR TITLE
[WIP] gh-81040: use sqlite3_stmt_readonly() to detect DML statements

### LINKED ISSUE #81040: sqlite3 dml statement detection does not account for CTEs
BPO | [36859](https://bugs.python.org/issue36859)
--- | :---
Nosy | @berkerpeksag, @animalize, @tirkarthi, @coleifer, @erlend-aasland
PRs | <li>python/cpython#13216</li><li>python/cpython#24492</li>
Files | <li>[36859.patch](https://bugs.python.org/file48319/36859.patch "Uploaded as text/plain at 2019-05-09.03:39:01 by @coleifer")</li><li>[36859-2.patch](https://bugs.python.org/file48320/36859-2.patch "Uploaded as text/plain at 2019-05-09.04:12:20 by @coleifer"): Working patch that retains backwards-compatibility.</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2019-05-08.23:08:24.384>
labels = ['3.8', 'type-bug', 'library']
title = 'sqlite3 dml statement detection does not account for CTEs'
updated_at = <Date 2021-08-11.21:02:17.366>
user = 'https://github.com/coleifer'
```

bugs.python.org fields:
```python
activity = <Date 2021-08-11.21:02:17.366>
actor = 'erlendaasland'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2019-05-08.23:08:24.384>
creator = 'coleifer'
dependencies = []
files = ['48319', '48320']
hgrepos = []
issue_num = 36859
keywords = ['patch']
message_count = 9.0
messages = ['341949', '341956', '341958', '341961', '341962', '386713', '386716', '386721', '399418']
nosy_count = 5.0
nosy_names = ['berker.peksag', 'malin', 'xtreak', 'coleifer', 'erlendaasland']
pr_nums = ['13216', '24492']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue36859'
versions = ['Python 3.8']
```

</p></details>

### PR DESCRIPTION
**Pull request by erlend-aasland**
Fixes #81040

### CODE REVIEW

**Comment by erlend-aasland** [comment]
Cherry-picked from #13216 and rebased onto master.

cc. @berkerpeksag @coleifer @serhiy-storchaka

File: Modules/_sqlite/statement.c [hunk 1]
```diff
@@ -48,13 +48,43 @@ typedef enum {
...
+                    continue;
+            }
+
+            is_dml = (PyOS_strnicmp(p, "begin", 5) &&
+                      PyOS_strnicmp(p, "create", 6) &&
+                      PyOS_strnicmp(p, "drop", 4) &&
+                      PyOS_strnicmp(p, "alter", 5) &&
+                      PyOS_strnicmp(p, "analyze", 7) &&
+                      PyOS_strnicmp(p, "reindex", 7) &&
+                      PyOS_strnicmp(p, "vacuum", 6));
```
Comment (erlend-aasland) [inline Modules/_sqlite/statement.c:75]: I don't think we need all of those.

- `sqlite3_stmt_readonly()` with `ALTER` statement returns `0`
- `sqlite3_stmt_readonly()` with `VACUUM` statement returns `0`
- `sqlite3_stmt_readonly()` with `ANALYZE` statement returns `0`
- `sqlite3_stmt_readonly()` with `DROP` statement returns `0`
- `sqlite3_stmt_readonly()` with `CREATE` statement that actually creates a new item returns `0`
- `sqlite3_stmt_readonly()` with `CREATE` statement that resolves to a no-op returns `1`

**Review by MaxwellDupre — CHANGES_REQUESTED** [review]
Ran 18 tests in 0.481s
OK
Other Sqlite tests also passed. Looks ok.
Need to add to Docs, not just NEWS otherwise users will not be aware.

**Comment by erlend-aasland** [comment]
I'm not sure this is the correct way to address to the problems in the linked issue. Closing.

**Pull request closed without merging**